pkg/provider/azure_vmss_cache.go

```diff
@@ -322,43 +353,120 @@ func (ss *ScaleSet) newAvailabilitySetNodesCache() (*azcache.TimedCache, error)
 			return nil, err
 		}
 
-		localCache := availabilitySetNodeEntry{
-			vmNames:   vmNames,
-			nodeNames: nodeNames,
-			vms:       vmList,
+		localCache := nonVmssUniformNodesEntry{
+			vmssFlexVMNodeNames:   vmssFlexVMNodeNames,
+			vmssFlexVMProviderIDs: vmssFlexVMProviderIDs,
+			avSetVMNodeNames:      avSetVMNodeNames,
+			avSetVMProviderIDs:    avSetVMProviderIDs,
+			clusterNodeNames:      nodeNames,
 		}
 
 		return localCache, nil
 	}
 
-	if ss.Config.AvailabilitySetNodesCacheTTLInSeconds == 0 {
-		ss.Config.AvailabilitySetNodesCacheTTLInSeconds = consts.AvailabilitySetNodesCacheTTLDefaultInSeconds
+	if ss.Config.NonVmssUniformNodesCacheTTLInSeconds == 0 {
+		ss.Config.NonVmssUniformNodesCacheTTLInSeconds = consts.NonVmssUniformNodesCacheTTLDefaultInSeconds
 	}
-	return azcache.NewTimedcache(time.Duration(ss.Config.AvailabilitySetNodesCacheTTLInSeconds)*time.Second, getter)
+	return azcache.NewTimedcache(time.Duration(ss.Config.NonVmssUniformNodesCacheTTLInSeconds)*time.Second, getter)
 }
 
-func (ss *ScaleSet) isNodeManagedByAvailabilitySet(nodeName string, crt azcache.AzureCacheReadType) (bool, error) {
-	// Assume all nodes are managed by VMSS when DisableAvailabilitySetNodes is enabled.
+func (ss *ScaleSet) getVMManagementTypeByNodeName(nodeName string, crt azcache.AzureCacheReadType) (VMManagementType, error) {
 	if ss.DisableAvailabilitySetNodes {
-		klog.V(6).Infof("Assuming node %q is managed by VMSS since DisableAvailabilitySetNodes is set to true", nodeName)
-		return false, nil
+		return ManagedByVmssUniform, nil
```

Comment:
this is actually not correct, `DisableAvailabilitySetNodes` means current cluster does not have avset (save vmas list operation), while it may have vmss uniform or vmss flex.

Reply:
Yes and No. The name DisableAvailabilitySetNodes is quite missleading. The starting point to have this flag is just to reduce API calls by not checking node types. Below is the detailed explanation:

- When CCM start to support AKS, all the nodes are VMSS Uniform nodes, so there is no need to distinguish node type at that time. 
- Later when CCM needs to support CAPZ, Uniform nodes and AvSet nodes are used together, so CCM added the availabilitySetNodesCache to distinguish them. More specifically, it added an availabilitySetNodesCache by using the VM List API to batch query all AvSet  (including standalone VMs) VM names in the resource group. After this, CCM does a check for the node type before each function execution. The side effect is, CCM needs to refresh the availabilitySetNodesCache periodically, which is an  additional cost. 
- Considering AKS only contains Uniform nodes, DisableAvailabilitySetNodes flag is added to bypass the node type check to reduce the unnecessary API call.
- Now, to support VMSS Flex, we should make the CCM to have backward compatilibity and does not affect the performance of existing execution. So, here if DisableAvailabilitySetNodes is set to true, we will not check the node type and assume all nodes are Uniform nodes, which is excactly the behavior before VMSS Flex is supported. In this way, the existing AKS cluster's performance using Uniform will not be affected.